solr/core/src/java/org/apache/solr/handler/admin/ZkRead.java

```diff
@@ -0,0 +1,117 @@
+import java.util.List;
+import java.util.Map;
+
+import org.apache.solr.api.Command;
+import org.apache.solr.api.EndPoint;
+import org.apache.solr.client.solrj.SolrRequest;
+import org.apache.solr.client.solrj.impl.BinaryResponseParser;
+import org.apache.solr.common.MapWriter;
+import org.apache.solr.common.params.CommonParams;
+import org.apache.solr.common.params.MapSolrParams;
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.common.util.ContentStreamBase;
+import org.apache.solr.common.util.Utils;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.apache.zookeeper.data.Stat;
+
+import static org.apache.solr.common.params.CommonParams.OMIT_HEADER;
+import static org.apache.solr.common.params.CommonParams.WT;
+import static org.apache.solr.response.RawResponseWriter.CONTENT;
+import static org.apache.solr.security.PermissionNameProvider.Name.COLL_READ_PERM;
+
+/**Exposes the content of the Zookeeper
+ * This is an expert feature that exposes the data inside the back end zookeeper.This API may change or
+ * be removed in future versions.
+ * This is not a public API. The data that is returned is not guaranteed to remain same
+ * across releases, as the data stored in Zookeeper may change from time to time.
+ */
+@EndPoint(path = "/cluster/zk/*",
+    method = SolrRequest.METHOD.GET,
+    permission = COLL_READ_PERM)
+public class ZkRead {
+  private final CoreContainer coreContainer;
+
+  public ZkRead(CoreContainer coreContainer) {
+    this.coreContainer = coreContainer;
+  }
+
+  @Command
+  public void get(SolrQueryRequest req, SolrQueryResponse rsp) {
+    String path = req.getPathTemplateValues().get("*");
+    if (path == null || path.isEmpty()) path = "/";
+    byte[] d = null;
+    try {
+      List<String> l = coreContainer.getZkController().getZkClient().getChildren(path, null, false);
+      if (l != null && !l.isEmpty()) {
+        String prefix = path.endsWith("/") ? path : path + "/";
+
+        rsp.add(path, (MapWriter) ew -> {
+          for (String s : l) {
+            try {
+              Stat stat = coreContainer.getZkController().getZkClient().exists(prefix + s, null, false);
+              ew.put(s, (MapWriter) ew1 -> {
+                ew1.put("version", stat.getVersion());
+                ew1.put("aversion", stat.getAversion());
+                ew1.put("children", stat.getNumChildren());
+                ew1.put("ctime", stat.getCtime());
+                ew1.put("cversion", stat.getCversion());
+                ew1.put("czxid", stat.getCzxid());
+                ew1.put("ephemeralOwner", stat.getEphemeralOwner());
+                ew1.put("mtime", stat.getMtime());
+                ew1.put("mzxid", stat.getMzxid());
+                ew1.put("pzxid", stat.getPzxid());
+                ew1.put("dataLength", stat.getDataLength());
+              });
+            } catch (Exception e) {
```

Comment:
Why are you catching `Exception` here? instead of a subclass? Also, `exists` throws interrupted exception, you should interrupt (and break)

Reply:
does it even matter? 

Why interrupt the thread when this thread is not doing anything at all. This is the `qtp` thread. Interrupting it makes no sense

Reply:
This try/catch is inside a loop that makes requests to ZooKeeper. So you are saying that if the thread is interruped (shutdown, or in tests for example) the plan is to swallow that interruption and continue processing (doing requests to ZooKeeper for every child this zkpath has, with sleeps/retries), and the intention of that is to add an "exception" to one (just one) of the elements in the tree in the response?
Regarding other exceptions, lets say you are getting back a SessionExpiredException, the plan is to continue iterating over all the child docs and continue adding `node -> KeeperException.SessionExpiredException`?
I think the proper handling of most these exceptions is to stop processing and return an error. If you get a NoNodeException, then yes, that could be added in the response if you want, but doing a blind `catch (Exception e)` is a bad idea.

Reply:
Please take a look at the latest

* NoNodeException returns with a `404`
* Other Exceptions return with `500`

This is supposed to just fail fast if there is a problem. It does not need to be any more sophisticated than that

Reply:
This is still a problem in your latest code

Reply:
I fail to see the problem. Could you please explain

Reply:
must be fixed now

Reply:
Is not yet fixed in the list case

Reply:
```
  public void list(SolrQueryRequest req, SolrQueryResponse rsp) {
      String path = req.getPathTemplateValues().get("*");
      if (path == null || path.isEmpty()) path = "/";
      try {
        List<String> l = coreContainer.getZkController().getZkClient().getChildren(path, null, false);
        String prefix = path.endsWith("/") ? path : path + "/";
        rsp.add(path, (MapWriter) ew -> {
          for (String s : l) {
            Stat stat = null;
            try {
              stat = coreContainer.getZkController().getZkClient().exists(prefix + s, null, false);
            } catch (Exception e) {
              throw new RuntimeException(e);
            }
            printStat(ew, s, stat);
          }
        });
      } catch (KeeperException.NoNodeException e) {
        throw new SolrException(SolrException.ErrorCode.NOT_FOUND, "No such node :"+ path);
      } catch (Exception e) {
        rsp.add(CONTENT, new ContentStreamBase.StringStream(Utils.toJSONString(Collections.singletonMap("error", e.getMessage()))));
      } finally {
        RequestHandlerUtils.addExperimentalFormatWarning(rsp);
      }
    }
```

If there is an exception you will see a 500 error in this code. Am I missing something?